### PR TITLE
[Lint] One parameter/argument per line for C++ code

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

It's really annoying to deal with parameter/argument conflicts. This is even frustrating when we merge code from the community to Ant's internal code base with hundreds of conflicts caused by parameters/arguments.

In this PR, I updated the clang-format style to make parameters/arguments stay on different lines if they can't fit into a single line.

There are several benefits:

* Conflict resolving is easier.
* Less potential human mistakes when resolving conflicts.
* Git history and Git blame are more straightforward.
* Better readability.
* Align with the new Python format style.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by raulchen** [comment]
I'm strongly in favor of putting one argument per line. The benefit is that this makes the git history very clear. E.g., you can easily find which PR introduced a given parameter. 
I used to comment this issue repeatedly. Glad that clang can automatically do this. 

@ericl @scv119 any comments?

**Comment by ericl** [comment]
Hmm, personally I think the ugliness is a bit much if the problem that could be solved in other ways. Maybe this is a sign we need narrower / cleaner interfaces in Ray for bits that need to be extended, or more components should be upstreamed?

**Comment by scv119** [comment]
cc @iycheng @mwtian 
personally I like the old formatting as it makes code more compact, but I not have a strong opinion here. Also agree with @ericl here it might mean the interfaces need to be narrowed.

**Comment by iycheng** [comment]
I think the original one seems better. Not sure, but for things like this:
```
  return ray::internal::Serializer::Deserialize<std::shared_ptr<T>>(
      packed_object->data(), packed_object->size());
```
to be formatted like this:
```
  return ray::internal::Serializer::Deserialize<std::shared_ptr<T>>(
      packed_object->data(),
      packed_object->size());
```

seems really wired.

I checked gRPC and tensorflow and they are not using this one and I'm not sure whether they have similar issues like the one described or not. It seems like other things are not working well, like what Eric mentioned.

I'll vote readability over the things in the descriptions.

**Comment by raulchen** [comment]
Regarding readability and aesthetics, I agree that some changes in this PR are not ideal. IIRC, Python's rule is that if all parameters can fit into a single line, put them in a single line, otherwise put each in one line. If we can adopt the same rule, I think this PR can improve code readability. And the amount of changes should be much less. @iycheng will this address your concern? 

Regarding code conflicts, we are indeed reviewing all existing diffs and planing to upstream as many existing diffs as possible throughout this year. If we could improve this format issues, it would greatly facilitate the process. 

Last but not least, I'd like to emphasize the benefit of clean git history again. I'v seen such cases many times. It was very inconvenient to git blame the real commit.

**Comment by iycheng** [comment]
I'm not 100% sure about the rules there. But I think if in the end, for functions which has more than 5 parameters, we can put them line-by-line maybe it's ok.
Also, I don't think cpp's style has to be aligned with python's style. They are different languages.

**Comment by jjyao** [comment]
Just my opinion: the python rule sounds good to me (and the clean git blame).

Could we apply python rule and update the PR so people can see how it looks in real?

Something I came across: https://softwareengineering.stackexchange.com/questions/88485/is-it-a-bad-idea-to-list-every-function-method-argument-on-a-new-line-and-why

**Comment by scv119** [comment]
> Regarding readability and aesthetics, I agree that some changes in this PR are not ideal. IIRC, Python's rule is that if all parameters can fit into a single line, put them in a single line, otherwise put each in one line. If we can adopt the same rule, I think this PR can improve code readability. 

I think this would be ideal.

File: .clang-format [hunk 1]
```diff
@@ -3,3 +3,11 @@ ColumnLimit: 90
 DerivePointerAlignment: false
 IndentCaseLabels: false
 PointerAlignment: Right
+BinPackArguments: false
+BinPackParameters: false
+AllowAllArgumentsOnNextLine: false
+AllowAllParametersOfDeclarationOnNextLine: false
+InsertTrailingCommas: Wrapped
```
Comment (mwtian) [inline .clang-format:10]: This option seems only useful for Javascript? https://clang.llvm.org/docs/ClangFormatStyleOptions.html

File: .clang-format [hunk 2]
```diff
@@ -3,3 +3,11 @@ ColumnLimit: 90
...
+AlignAfterOpenBracket: AlwaysBreak
+AllowAllConstructorInitializersOnNextLine: true
```
Comment (mwtian) [inline .clang-format:12]: Is this consistent, with `AllowAllConstructorInitializersOnNextLine` to `true` and other `*OnNextLine` settings `false`? `folly` seems to set `BinPack*` to false and `*OnNextLine` to true.

**Review by mwtian — COMMENTED** [review]
I'm ok as long as it is consistent across the code base for most C++ constructs, and similar to a fraction of other C++ OSS projects.

One disadvantage of the new style is that when reading code, more vertical screen estate will be used for the same amount of information.

File: .clang-format (hunk 1 above)
Comment (kfstorm) [reply .clang-format:10]: Thanks. I'll revisit all configurations.

**Comment by kfstorm** [comment]
Everyone, I've revisited the configurations and simplifed them to avoid unnecessary changes.

Simple stats for comparison:

The original style and stats:

```
BinPackArguments: false
BinPackParameters: false
AllowAllArgumentsOnNextLine: false
AllowAllParametersOfDeclarationOnNextLine: false
InsertTrailingCommas: Wrapped
AlignAfterOpenBracket: AlwaysBreak
AllowAllConstructorInitializersOnNextLine: true
ConstructorInitializerAllOnOneLineOrOnePerLine: true

449 files changed, 22015 insertions(+), 12225 deletions(-)
```

The new style and stats:

```
BinPackArguments: false
BinPackParameters: false

377 files changed, 9807 insertions(+), 5166 deletions(-)
```

**Comment by kfstorm** [comment]
> I'm not 100% sure about the rules there. But I think if in the end, for functions which have more than 5 parameters, we can put them line-by-line maybe it's ok. I also think for this case, we probably should do it like:
> 
> ```
> fun_call(
>   param1, /* name1 */
>   param2, /* name2 */
>   param3, /* name3 */
>   param4, /* name4 */
>   param5, /* name5 */
>   param6 /* name6 */
> )
> ```
> 
> Also, I don't think cpp's style has to be aligned with python's style. They are different languages.

I can't find a configuration to have different formatting behaviors based on the count of parameters.

**Comment by mwtian** [comment]
Nice that we can trim down the formatting rule change and code diff. There are other cost to the change, e.g. deviation from the base Google style, disruption to existing PRs etc. But LGTM if there is consensus on this.

**Review by raulchen — APPROVED** [review]
LGTM after trimming down the rules.

**Comment by kfstorm** [comment]
@ericl @scv119 @iycheng Are you OK with the new changes? If yes, we can then move on to setup a timeline for migration and send reminders to open pull requests just like what we did for Python migration to black (https://github.com/ray-project/ray/issues/21311).

**Review by ericl — APPROVED** [review]
Lgtm

**Review by scv119 — APPROVED** [review]

**Comment by jjyao** [comment]
Can we only send out comments to PRs that touches c++?

**Comment by kfstorm** [comment]
> Can we only send out comments to PRs that touches c++?

Sure, unless there's no way to find out what files are changed in a PR.

**Pull request merged**